main.go

```diff
@@ -81,21 +98,45 @@ func main() {
 			})
 		}
 
+		messagesMutex.Lock()
 		messages = append(messages, body.Message)
+		messagesMutex.Unlock()
 
-		destinations := topology[n.ID()]
+		go func(destinations []string, body BroadcastRequestBody) {
+			var successfulSentDestinations []string
+			var successfulSentDestinationsMutex sync.Mutex
+			// Send messages to the node's destinations
+			// Until we got none left
+			for len(destinations) != len(successfulSentDestinations) {
+				for _, node := range destinations {
+					n.RPC(node, body, func(msg maelstrom.Message) error {
+						var broadcastOkBody BroadcastOkResponseBody
 
-		if destinations == nil {
-			return n.Reply(msg, BroadcastResponseBody{
-				Type: "broadcast_ok",
-			})
-		}
+						if err := json.Unmarshal(msg.Body, &broadcastOkBody); err != nil {
+							return err
+						}
 
-		for _, node := range destinations {
-			go func(node string) {
-				n.Send(node, body)
-			}(node)
-		}
+						if broadcastOkBody.Type != "broadcast_ok" {
+							return fmt.Errorf("expeted type broadcast_ok, got %s", broadcastOkBody.Type)
+						}
+
+						log.Printf("from %s, to %s, sent %v", n.ID(), node, body)
+
+						successfulSentDestinationsMutex.Lock()
+						defer successfulSentDestinationsMutex.Unlock()
+
+						// Check if we already marked the destination node as successfully sent
+						if !slices.Contains(successfulSentDestinations, node) {
+							successfulSentDestinations = append(successfulSentDestinations, node)
+						}
+
+						return nil
+					})
+				}
+
+				time.Sleep(1 * time.Second)
```

Comment:
Should we decide to tackle a higher scale system, e.g. as in detailed in [challenge #3d](https://fly.io/dist-sys/3d/) (Efficient Broadcast, Part I), this timer would have to be increased to >=2s, as it otherwise overwhelms the system